src/ComplexSelector.js

```diff
@@ -0,0 +1,128 @@
+import reduce from 'lodash/reduce';
+import split from 'lodash/split';
+
+export default class ComplexSelector {
+  constructor(buildPredicate, findWhereUnwrapped, childrenOfNode) {
+    this.buildPredicate = buildPredicate;
+    this.findWhereUnwrapped = findWhereUnwrapped;
+    this.childrenOfNode = childrenOfNode;
+  }
+
+  getSelectors(selector) {
+    const cleaned = selector.replace(/\s{2,}/g, ' ');
+    const selectors = split(cleaned, ' ');
+    return reduce(selectors, (memo, sel) => {
+      if (sel === '+' || sel === '~') {
+        const temp = memo.pop();
+        memo.push(sel, temp);
+        return memo;
+      }
+
+      memo.push(sel);
+      return memo;
+    }, []);
+  }
+
+  handleSelectors(selectors, wrapper) {
+    const recurseSelector = (offset, fn, pre) => {
+      const predicate = pre || this.buildPredicate(selectors[offset]);
+      const nextWrapper = this.findWhereUnwrapped(wrapper, predicate, fn);
+      const nextSelectors = selectors.slice(offset + 1);
+      return this.handleSelectors(nextSelectors, nextWrapper);
+    };
+
+    const buildSiblingPredicate = (first, second) => {
+      const firstPredicate = this.buildPredicate(first);
+      const secondPredicate = this.buildPredicate(second);
+
+      return (child) => {
+        if (firstPredicate(child)) {
+          return (sibling) => secondPredicate(sibling);
+        }
+
+        return false;
+      };
+    };
+
+    let { predicate, selectSiblings } = {};
```

Comment:
I usually do this when defining multiple variables to undefined, but `let predicate; let selectSiblings;` is also fine